config/src/utils.rs

```diff
@@ -44,6 +101,110 @@ fn get_random_port() -> ::std::io::Result<u16> {
     Ok(addr.port())
 }
 
+fn lock_path() -> String {
+    format!(
+        "/tmp/aptos-port-counter.{}.lock",
+        &NEXTEST_RUN_ID.clone().unwrap()
+    )
+}
+
+fn counter_path() -> String {
+    format!(
+        "/tmp/aptos-port-counter.{}",
+        &NEXTEST_RUN_ID.clone().unwrap()
+    )
+}
+
+/// We use the filesystem to bind to unique ports for cargo nextest.
+/// cargo nextest runs tests concurrently in different processes. We have observed that using
+/// a simple bind(0) results in flaky tests when nodes are restarted within tests; likely the OS
+/// is prioritizing recently released ports.
+fn get_unique_port() -> u16 {
+    let mut port_counter_files = open_counter_file();
+    let global_counter = match port_counter_files.counter_file.read_u16::<BigEndian>() {
+        Ok(counter) => {
+            if counter as usize >= PORT_VECTOR.len() {
+                0
+            } else {
+                counter
+            }
+        }
+        Err(_) => {
+            warn!(
+                "Unable to read port counter from file {}, starting from 0",
+                counter_path()
+            );
+            0
+        }
+    };
+    let (port, updated_counter) = bind_port_from_counter(global_counter);
```

Comment:
this feels a race condition, I guess we just retried until succeed if the file is written out of order?

Reply:
This is protected by the lock_file in open_counter_file. Only one process will create the lock_file, other processes will wait until it's removed.